### PR TITLE
Emit metrics when queries exceed query limits

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR introduces a `cortex_queries_exceeded_limits_total` counter that is incremented when a query exceeds the max series, max chunks or max chunk bytes limits.

The counter is labelled with the name of the limit (eg. `max-fetched-chunks-per-query`).

Note to reviewers: unlike the similar limit-related metrics emitted by the distributor, I haven't broken these down by tenant to keep cardinality manageable, but open to feedback if you think that would be valuable.

I'll put together a follow-up PR to add this to a dashboard in the mixin once this PR is merged.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated
- [n/a] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/limiter/query_limiter.go [hunk 1]
```diff
@@ -75,7 +80,7 @@ func QueryLimiterFromContextWithFallback(ctx context.Context) *QueryLimiter {
 	ql, ok := ctx.Value(ctxKey).(*QueryLimiter)
 	if !ok {
 		// If there's no limiter return a new unlimited limiter as a fallback
-		ql = NewQueryLimiter(0, 0, 0)
+		ql = NewQueryLimiter(0, 0, 0, nil)
```
Comment (pracucci) [inline pkg/util/limiter/query_limiter.go:83]: This means the metrics struct may be nil, but the rest of the code doesn't handle that case.

File: pkg/util/limiter/query_limiter.go [hunk 2]
```diff
@@ -93,7 +98,11 @@ func (ql *QueryLimiter) AddSeries(seriesLabels []mimirpb.LabelAdapter) error {
 
 	ql.uniqueSeries[fingerprint] = struct{}{}
 	if len(ql.uniqueSeries) > ql.maxSeriesPerQuery {
-		// Format error with max limit
+		if len(ql.uniqueSeries) == ql.maxSeriesPerQuery+1 {
+			// If we've just exceeded the limit for the first time for this query, increment the failed query metric.
+			ql.queryMetrics.QueriesExceededLimits.WithLabelValues("max-fetched-series-per-query").Inc()
```
Comment (pracucci) [inline pkg/util/limiter/query_limiter.go:103]: [nit] I'm wondering if really need the `-per-query` suffix here and in other limits too. Given this metric tracks the limits per query "by definition", we could probably omit the suffix.

File: pkg/querier/stats/query_metrics.go [hunk 3]
```diff
@@ -33,5 +36,10 @@ func NewQueryChunkMetrics(reg prometheus.Registerer) *QueryChunkMetrics {
 			Name:      "distributor_query_ingester_chunks_total",
 			Help:      "Number of chunks received from ingesters at query time, including any chunks discarded as duplicates.",
 		}),
+		QueriesExceededLimits: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "cortex",
+			Name:      "queries_exceeded_limits_total",
```
Comment (pracucci) [inline pkg/querier/stats/query_metrics.go:41]: [nit] Should we prefix with "cortex_querier" to keep it consistent with other querier metrics? These limits are enforced only in the querier, right? I don't feel strong about this comment.

```suggestion
			Name:      "querier_queries_exceeded_limits_total",
```

**Review by pracucci — APPROVED** [review]
The metric itself LGTM but I'm missing how it's going to be queried. I think querying the absolute value (or rate) is not much useful, while it could be more useful to compute the % of queries which get limited. To compute the % of limited queries, what's the metric are you going to use to get the total number of queries?

Please look at the comment about `nil` metrics struct.

File: pkg/util/limiter/query_limiter.go (hunk 1 above)
Comment (charleskorn) [reply pkg/util/limiter/query_limiter.go:83]: In the case where the limits are 0, the checks are skipped and the metrics struct isn't used.

File: pkg/util/limiter/query_limiter.go (hunk 2 above)
Comment (charleskorn) [reply pkg/util/limiter/query_limiter.go:103]: I wen't with the `-per-query` suffix to keep the name the same as the CLI option (and therefore make it easy and unambiguous to find in the docs). I'm happy either way though.

File: pkg/querier/stats/query_metrics.go (hunk 3 above)
Comment (charleskorn) [reply pkg/querier/stats/query_metrics.go]: Changed in b917f19

**Comment by charleskorn** [comment]
>To compute the % of limited queries, what's the metric are you going to use to get the total number of queries?

I'm thinking I'd use the same metric we use on the dashboards - something like `cortex_querier_request_duration_seconds_count{route=~"(prometheus|api_prom)_api_v1_query(_range)?"})`. Does that make sense? Or is there an issue with using that?

**Comment by pracucci** [comment]
> I'm thinking I'd use the same metric we use on the dashboards - something like cortex_querier_request_duration_seconds_count{route=~"(prometheus|api_prom)_api_v1_query(_range)?"}). Does that make sense?

Ok that should work, given it's tracked in the querier, so for each partial query like the limit one.

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Review by dimitarvdimitrov — COMMENTED** [review]
We also can add the new metrics to a panel in the "Queries" dashboard.

**Comment by charleskorn** [comment]
> We also can add the new metrics to a panel in the "Queries" dashboard.

Yep, I was thinking of doing that as a separate PR once we've deployed this somewhere  and I've had a chance to use it, just to see what the most useful way to visualise this is (eg. absolute rate of rejected queries or rejected queries as a proportion of all query requests)

**Pull request merged**